kips/KIP01 - Prioritize Kolibri Liquidity Pool for Liquidations.md

```diff
@@ -0,0 +1,32 @@
+# Kolibri Improvement Proposal #01
+
+### Preface
+Hello everyone :)  
+I am honored to present the first Kolibri Improvement Proposal. I feel like Satoshi in some sense right now.  
+Kidding aside, I welcome and encourage everyone to participate in this and other proposals.  
+Lets not forget that as kDAO holders and as a community, we are in the control of the future and success of Kolibri.  
+I should also thank the Kolibri dev team for their hardwork and support.
+
+### Summary
+An onchain executable procedure to priorotize the Kolibri Liquidity Pool for oven liquidations.
+
+### Motivation
+The Kolibri Liquidity Pool (KLP) is an important and integral part of the kUSD algorithmic stable coin.  
+It serves at least the following purposes :
+1. Makes sure that big ovens that may not be liquidable by individuals, can be liquidated by the pooling the resources of individual kUSD investors.
+2. Provides a fair return for the investors who are willing to help the kUSD ecosystem stability. 
+
+
+The current implementation doesnt prioritize the KLP and individual investors compete for the liquidation of the under-collaterized ovens using high fees.  
+Such implementation is not healthy for the ecosystem and will discourage the pool investors and may lead to disintegration of KLP due to small or no returns of their investment into KLP.  
+In order to resolve this issue, I propose to give the KLP a X block priority for the liquidations.
+
+### Details
+In order to make an onchain executable procedure, I propose to make the Liquidation a 2 step process, Preliquidate and Liquidate.  
```

Comment:
Here's a quick high level view of how this works today. 

![Kolibri Diagram](https://user-images.githubusercontent.com/18271723/128254758-ab2f1b69-a695-4b53-a06d-ee4463325cb8.jpeg)


I think at a technical level to implement your proposal:
- We'd need to design a new oven contract, which holds a these new bits of information (at the very least, `isPreliqudate`. I suspect we may need more, see below)
- We'd need to update `OvenFactory` to issue new `Oven` contracts
 - New APIs to accommodate pre-liquidate should be created in `Minter` and we'll need to change the minter contract. 
 - A new `OvenProxy` will need to shim the new ovens to the new minter.
 
 So we'd end up with:
 - A set of Oven v1s (these will always liquidate like normal)
 - A set of Oven v2s
 - Oven Proxy v1 and Oven Proxy V2, which shim calls to the Minter
 - Minter V2 (Minter V1 is completely deprecated)
 

Note: I've architected this system but I welcome debate on if there's a better technical design - I'm not an infallible engineer.

Reply:
I totally understand my proposal needs updating contracts and adds complexity, I am not a blockchain engineer so I dont comment on the design.
But if it is not worth implementing technically, probably only allowing the KLP to liquidate the ovens is the way to go.

Reply:
@keefertaylor wouldn't we be able to gate/manage this entirely at the minter? 

I feel like we might be able to actually hoist this into the existing system by upgrading the minter and rewriting the `%liquidate` function to call into a different contract that could manage state (or just manage the liquidation status itself in a map or something). 

Then from the different roles:
- Liquidity pool would be able to just liquidate underwater ovens no matter what (no complicated timer needed)
- Stability fund would be the same deal
- Arbitrary liquidators would just be added to an internal map of liquidation requests along with the block they requested, things *are not* liquidated, but if that same liquidator comes along and after 5 blocks calls `liquidate()` on the oven again, it's able to go through. 

Unless I'm missing something, I'm not sure there's anything else that needs updated but that one function. Benefits are minimal overhaul to the system, gives us the timeout mechanism we want to allow for the LP to liquidate things, and allows for individuals to liquidate as a last resort with a large economic incentive to keep on top of it.

Reply:
Additionally, if after 5 blocks pass the oven is no longer liquidatable, the `%liquidate` entrypoint should just clean things up.